chore: Streamline auth take two

EyeCantCU opened the pull request:
* Automate token generation
* Use Octo STS policy for guarded repos

EyeCantCU committed:
fix(Makefile): Just use grep

Ubuntu ships an older version of yq that isn't as featureful. To ensure
the check just works, just use grep

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
Reapply "chore: Streamline auth flow (#55580)" (#56092)

This reverts commit 9033860e8dc93668f4e653cd963d2afa43ea806c.

EyeCantCU committed:
chore(pipelines/auth): Prefer Octo STS over local tokens

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(Makefile): Skip token generation when auth'd Google SA is in use

When we can use a Google SA to authenticate with issuer.enforce.dev,
skip token generation (workstations)

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

stevebeattie reviewed (commented):
This is looking pretty nice, a couple of nits.

EyeCantCU committed:
fix(Makefile): Filter for pipelines in compiled melange manifests

Use melange compile to output a compiled package manifest, inclusive of
all pipelines, that we can parse with jq and use for auth

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(Makefile): Revert auth with Google SA for now

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU commented on Makefile:
```diff
@@ -100,14 +104,52 @@ clean:
+	tmpf=$(shell mktemp) && \
+	chainctl auth octo-sts --identity=guarded-package-repos --scope=chainguard-dev > $${tmpf} && \
+	mkdir -p $(pkgname) && \
+	mv $${tmpf} ./$(pkgname)/.guarded-repo.token
```
I think this is good as is, but feel free to disagree and I can revisit. Every recipe that calls this has a trap that performs cleanup of the file after this target is called, regardless of whether or not chainctl succeeds. There isn't a great way to handle this in the target itself because it has been called by another consumer. If chainctl fails, that failure will be printed to stderr and will not be written to the file. If mkdir fails... we probably have worse things to worry about. We are still left with a file that contains whatever was written to stdout, but that's only possible if manually invoked, which I hope is a very very rare edge case after this

EyeCantCU commented:
I really like the idea of writing the compiled manifest to `compiled/packagename.json`... Let's do that. Less error prone, and we don't have to silence output for the compile target to get what we need

EyeCantCU committed:
fix(make/compile): Write compiled manifests to compiled subdir

- Write compiled manifests to compiled subdir rather than printing it
- Parse the compiled manifest rather than make output during token
  generation
- Restore output for compile make target

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(pipelines/auth): Prioritize local token over Octo-STS

If a Google SA is present, authentication will fail if it does not have
required permissions

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(pipelines/auth): Use package.srcdir for token discoverability

Cosmetic, this should always be /home/build

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(pipelines/auth): Remove token file from build env after use

Allow opt out via cleanup-token toggle. I imagine the cases where we
need to reuse are few and far between. The last pipeline can at least
still perform removal

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

stevebeattie commented on pipelines/auth/guarded-repo.yaml:
```diff
@@ -0,0 +1,43 @@
+  cleanup-token:
+    description: whether or not to remove token file after use
+    default: true
+
```
I like this but unfortunately, there are a few packages that have build pipelines that end up with multiple `uses: iamguarded/build-compat` steps, so either the `cleanup-token` option needs to be surfaced upwards into that pipeline.

stevebeattie reviewed (approved):
Thanks, this is working well for me.

We should probably have a followup to have `make clean` clear out the `compiled/` directory, and I would like to figure out a decent way to clean out the token before any build steps start, without breaking multi-pipeline builds.
